File: mobile/lib/protocol/disconnect_payload.dart
import 'dart:typed_data';

import 'frame.dart';
import 'payload_codec.dart';

const int disconnectReasonNormal = 0x00;
const int disconnectReasonAppClosing = 0x01;
const int disconnectReasonServerRestart = 0x02;
const int disconnectReasonIdleTimeout = 0x03;
const int disconnectReasonSecurity = 0x04;
const int disconnectReasonProtocolViolation = 0x05;

class DisconnectPayload {
  const DisconnectPayload({required this.reason});

  final int reason;

  @override
  bool operator ==(Object other) =>
      other is DisconnectPayload && reason == other.reason;

  @override
  int get hashCode => reason.hashCode;
}

class DisconnectPayloadCodec {
  static Uint8List encode(DisconnectPayload payload) {
    if (!_isKnownReason(payload.reason)) {
      throw const ProtocolException('Unknown DISCONNECT reason.');
    }

    final writer = PayloadWriter()..writeUInt8(payload.reason);
    return writer.toBytes();
  }

  static DisconnectPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final reason = reader.readUInt8();
    reader.expectEnd();

    if (!_isKnownReason(reason)) {
      throw const ProtocolException('Unknown DISCONNECT reason.');
    }

    return DisconnectPayload(reason: reason);
  }

  static bool _isKnownReason(int reason) => reason <= disconnectReasonProtocolViolation;
}

File: desktop/Protocol/DisconnectPayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record DisconnectPayload(byte Reason);

public static class DisconnectPayloadCodec
{
    public const byte ReasonNormal = 0x00;
    public const byte ReasonAppClosing = 0x01;
    public const byte ReasonServerRestart = 0x02;
    public const byte ReasonIdleTimeout = 0x03;
    public const byte ReasonSecurity = 0x04;
    public const byte ReasonProtocolViolation = 0x05;

    public static byte[] Encode(DisconnectPayload payload)
    {
        if (!IsKnownReason(payload.Reason))
            throw new ProtocolException("Unknown DISCONNECT reason.");

        var writer = new PayloadWriter();
        writer.WriteUInt8(payload.Reason);
        return writer.ToArray();
    }

    public static DisconnectPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var reason = reader.ReadUInt8();
        reader.ExpectEnd();

        if (!IsKnownReason(reason))
            throw new ProtocolException("Unknown DISCONNECT reason.");

        return new DisconnectPayload(reason);
    }

    private static bool IsKnownReason(byte reason) => reason <= ReasonProtocolViolation;
}
